src/libcore/char.rs

```diff
@@ -448,19 +459,15 @@ impl Iterator for EscapeUnicode {
                 Some('u')
             }
             EscapeUnicodeState::LeftBrace => {
-                let mut n = 0;
-                while (self.c as u32) >> (4 * (n + 1)) != 0 {
-                    n += 1;
-                }
-                self.state = EscapeUnicodeState::Value(n);
+                self.state = EscapeUnicodeState::Value;
                 Some('{')
             }
-            EscapeUnicodeState::Value(offset) => {
-                let c = from_digit(((self.c as u32) >> (offset * 4)) & 0xf, 16).unwrap();
-                if offset == 0 {
+            EscapeUnicodeState::Value => {
+                let c = from_digit(((self.c as u32) >> (self.offset * 4)) & 0xf, 16).unwrap();
```

Comment:
Yes it would be possible to simplify the code in `next` as you suggest:
```Rust
// here
let nibble = self.c >> 28;
self.c <<= 4;
if self.c == 0 { self.state = RightBrace }
from_digit(nibble)
```
There would still be the need to compute the remaining number of hex digits to be emitted in `size_hint`, but that would still remove the code duplication (and maybe help with the naming by making it more contextual).